interp/interp.go

```diff
@@ -799,26 +799,46 @@ func (r *Runner) stmtSync(ctx context.Context, st *syntax.Stmt) {
 	}
 }
 
-func (r *Runner) sub() *Runner {
+// Subshell makes a copy of the given Runner, suitable for use in parallel to
+// the original.  The copy will have the same parameters, environment
+// variables, etc, but they can all be modified without affecting the
+// original.
+//
+// Subshell is not safe to run from a different goroutine if r is in the
+// middle of Run().  Orchestrating this is left up to the caller; no locking
+// is performed.
+//
+// To replace e.g. stdin/out/err, do StdIO(r.stdin, r.stdout, r.stderr)(r) on
+// the copy.
+func (r *Runner) Subshell() *Runner {
 	// Keep in sync with the Runner type. Manually copy fields, to not copy
 	// sensitive ones like errgroup.Group, and to do deep copies of slices.
 	r2 := &Runner{
 		Env:         r.Env,
 		Dir:         r.Dir,
-		Params:      r.Params,
 		execHandler: r.execHandler,
 		openHandler: r.openHandler,
 		stdin:       r.stdin,
 		stdout:      r.stdout,
 		stderr:      r.stderr,
 		filename:    r.filename,
 		opts:        r.opts,
+		usedNew:     r.usedNew,
 
 		origStdout: r.origStdout, // used for process substitutions
 	}
+	r2.Params = append([]string(nil), r.Params...)
 	r2.Vars = make(map[string]expand.Variable, len(r.Vars))
 	for k, v := range r.Vars {
-		r2.Vars[k] = v
+		v2 := v
+		v2.List = append([]string(nil), v.List...)
+		if v.Map != nil {
+			v2.Map = make(map[string]string, len(v.Map))
```

Comment:
why do the nil check for Map, but not List? I think the compiler should be smart enough to not allocate for the base case of `len(x) == 0` in both cases.

Reply:
Or, if that assumption is wrong, we could add a little benchmark to track allocations. I don't think we're that far yet, but for that same reason, I think we should avoid extra code for optimizations we're not sure about. And I'm personally not sure about this one.

Reply:
`v2.List = append([]string(nil), v.List...)` is kind of neat.  If `v.List` is `nil`, then after this `append`, `v2.List` will also (still) be `nil`.  This is different from `v2.List = append([]string{}, v.List...)`, which will result in `v2.List` being a non-`nil` slice of length 0 (when `v.List` is `nil`).

```go
var x1, x2 []int
x1 = append([]int{}, x2...)
fmt.Printf("x1: %#v, is nil: %t\n", x1, x1 == nil)

x1 = nil
x1 = append([]int(nil), x2...)
fmt.Printf("x1: %#v, is nil: %t\n", x1, x1 == nil)
```
=>
```
x1: []int{}, is nil: false
x1: []int(nil), is nil: true
```
[(playground)](https://play.golang.org/p/oAB1PAD9QDw)

`make(map...)` on the other hand will always result in a non-`nil` map, even if it's of length 0.

```go
m1 := make(map[int]bool, 0)
fmt.Printf("m1: %#v, is nil: %t\n", m1, m1 == nil)
```
=>
```
m1: map[int]bool{}, is nil: false
```
(Same playground as above.)

I wasn't sure whether the rest of the code cared about `Map == nil` vs `len(Map) == 0`, so I erred on the side of identical semantics.  It wasn't really about optimization or allocations, just that if your source is `nil` and your copy is non-`nil` that seemed like a bad thing.

Reply:
Ah, gotcha. Then please leave a comment like `// Ensure that List and Map remain nil if they were nil`

Reply:
How about

```go
// Make deeper copies of List and Map, but ensure that they remain nil
// if they are nil in v.
```
?

Hmm.  About that `v2 := v`.  Should I do the same field-by-field copy for `v` that I do for `r`, for the same reason?  Or is `expand.Variable` sufficiently primitive to not need it?

Reply:
Your wording SGTM.

I don't think expand.Variable is going to change enough to warrant the same kind of defensive programming :)